### ISSUE TITLE
Organisations provisioning

### ISSUE THREAD

**Issue by schweikert** [grafana/grafana#12119]
Currently configuring organisations and LDAP group mapping to organisation roles in an automated way is quite difficult. If you start with no organisations and want to configure them in an automatic way, you need to:

1. Start grafana with no group mapping in ldap.toml
2. Call the grafana API multiple times to create the organisations and get their IDs
3. Generate a new ldap.toml with the IDs of the organisations
4. Restart grafana

Especially in a Kubernetes environment this becomes very difficult to engineer right. We want to be able to recreate the environment with all organisations in an automated way. It would be a lot easier if provisioning at startup-time of organisations would be possible.

Maybe grafana-cli might help there, so that you could create the orgs before starting up grafana. Also, it should be possible to specify fixed IDs, so that you can also provide a static ldap.toml file (that you could generate with an init container, for example).

**Comment by bergquist** [comment]
If your running Grafana in k8s It might be easier to run one pod per org instead?

**Comment by fessmage** [comment]
+1 vote for possibility of organisations and users provisioning

**Comment by Anonymous-Coward** [comment]
> If your running Grafana in k8s It might be easier to run one pod per org instead?

Not quite. The (IME) typical depoyment in k8s is via prometheus-operator. That deploys once per cluster, with a single grafana pod, and is not really friendly to customize to that level of detail as to add multiple grafanas, and have them preconfigured per org. But you may very well want to have multiple orgs in a cluster.

**Comment by olivierlemasle** [comment]
Hi, I'd be happy to work on this.

However, I have some questions:

- First of all, will such a PR be merged? Is it still [Grafana's plans](https://github.com/grafana/grafana/tree/3e5abe7/pkg#provisionable) to have everything "provisionable"?

- Organizations have currently no unique identifier except the `id` provided by the database. Should we add a new `uid` to organizations, such as what was done in f132e929c for alert notifiers (in the process of adding notifiers provisioning) or what I've done for playlists (in PR #19040 - not yet reviewed). In that case, I suppose we should make it possible to use the `uid` while provisioning dashboards, datasources and notifiers, and in LDAP configuration (to have dashboards, users, etc. created in the right organization).

- Should we do something with organization [_addresses_](https://github.com/grafana/grafana/blob/6d70545/pkg/models/org.go#L19)? These addresses do not appear in Grafana UI, and [API related to these fields](https://github.com/grafana/grafana/blob/6d70545c0785cb206992093feaed449311401a46/pkg/api/org.go#L121) is not documented. Is it more than a draft feature that landed in Grafana codebase?

**Comment by marefr** [comment]
@olivierlemasle we'll need to do some additional thinking regarding supporting "everything to be provisioned", for now organizations and playlists. That's main reason we haven't had possibility to review your playlist provisioning PR. It's on our agenda to discuss this soon so please have patience. Thanks

**Comment by Mario-Hofstaetter** [comment]
Is there any update on provisioning organisations?

I just tried setting the Default Main Org Home Dashboard on a newly provisioned grafana instance in docker and found no way of doing it, as I have to login with my `admin` , `star` my custom `Home` dashboard end set it as `Home Dashboard` in the `Preferences` of the Default Org.

I don't want to script the http API within docker just to set the homepage.

**Comment by philschleier** [comment]
I've been using basic org provisioning with name and id.

```
orgs:
- id: 1
  name: First Org
- id: 2
  name: Second Org
deleteOrgs:
- id: 3
```

So far I haven't had any issues, but might very well be breaking stuff elsewhere that I'm not aware of...

I can open a PR if that's of interest?

**Comment by andrewrisse** [comment]
> r

Did you put this configuration under grafana.ini or somewhere else?

**Comment by andrewrisse** [comment]
> I've been using basic org provisioning with name and id.
> 
> ```
> orgs:
> - id: 1
>   name: First Org
> - id: 2
>   name: Second Org
> deleteOrgs:
> - id: 3
> ```
> 
> So far I haven't had any issues, but might very well be breaking stuff elsewhere that I'm not aware of...
> 
> I can open a PR if that's of interest?

Did you put this configuration under grafana.ini or somewhere else?

**Comment by tanandy** [comment]
Hi, since its a most wanted feature, could we have some feedback about it please ?

**Comment by Mario-Hofstaetter** [comment]
@tanandy and anyone subscribed to this thread:

As of 2022, it seems using orgs is discouraged:
https://grafana.com/blog/2022/03/14/how-to-best-organize-your-teams-and-resources-in-grafana

Money quotes:

> However, we rarely recommend Orgs as a way to separate teams, because they lack the flexibility of Folders and the true isolation of Instances and Stacks.

> We also plan to improve Grafana’s provisioning, APIs, and as-code functionality, to make it easier to manage resources between Instances.
> [...]
> **We do not plan to invest heavily in Orgs, because Folders and Teams can provide similar functionality with greater flexibility.**

What I take from this is that I do not expect this feature to be implemented anymore and will continue to avoid the need for orgs / provisioning them.
If multiple isolated grafana instances are necessary, I will deploy them in our stack each with the complete configuration (unique settings as necessary, etc.)

**Comment by burkempers** [comment]
I would still like to see this feature.  I say this due to the anonymous access setup inside of Grafana.  Grafana only allows anonymous access per OrgId and not teams.  This function is still necessary to have different dashboards for internal use versus external customer use.  So long as anonymous access is configured by OrgId there still should be a way to provision additional orgs.

**Comment by double-yaya** [comment]
so, how to add orgs before grafana starting

**Comment by lek-x** [comment]
any updates?

**Comment by rkostyantyn** [comment]
Also interested in this feature (orgs provisioning)

**Comment by dwt** [comment]
Is there any update on this? The linked blog post talks about teams being the future - but it seems that there is not even a bug / pull request / documentation on how to provision teams.

**Comment by ppaulweber** [comment]
Any updates on this feature request?

**Comment by spatterIight** [comment]
sad !

**Comment by EZ4BRUCE** [comment]
any updates ?

**Comment by Nine-H** [comment]
> 2018
> In v5.0 we decided to improve this experience by adding a new active provisioning system that uses config files. This will make GitOps more natural as data sources and dashboards can be defined via files that can be version controlled. We hope to extend this system to later add support for users, orgs and alerts as well.
> https://grafana.com/docs/grafana/latest/administration/provisioning/

> 2022
> We rarely recommend Orgs as a way to separate teams
> https://grafana.com/blog/2022/03/14/how-to-best-organize-your-teams-and-resources-in-grafana

> 2023
> Still, I find it sad to refuse this contribution. This is adding value.
> https://github.com/grafana/grafana/pull/78253

The situation is that provisioning orgs through the provisioning directory is unimplemented, has been left unimplemented for more than five and a half years, the advice offered by grafana is to use a feature that is completely useless for isolating resources without an enterprise license, and community patches that implement the functionality have been shot down.

> FYI, generally speaking, we are not investing in organizations and are working on improving Grafana Teams as an alternative.
> https://github.com/grafana/grafana/pull/77683

I doubt the proposed improvements to teams involve opening RBAC to OSS users to support small scale self managed deployments.

You have to love open core software and the open core business model.

**Comment by sathieu** [comment]
Proposed (but refused) implementation: https://github.com/grafana/grafana/pull/78253

**Comment by zorek187** [comment]
Seems like provisioning organisations is a workaround to Grafana Enterprise's RBAC feature which isn't likely to be added to OSS..
Still would love to get org provisioning in OSS

**Comment by yashikapathy28** [comment]
Hi,
I wanted to check if there are any updates on this issue.

We are trying to manage this as part of provisioning, specifically when configuring orgId for datasources and dashboards. In our Kubernetes deployment, setting an orgId other than the default (Org 1) causes Grafana to go into a CrashLoopBackOff, which is not the expected behavior.

From our understanding, organization creation is currently supported only via Admin API calls, which occur after Grafana is up and running—not during startup via provisioning. This appears to conflict with the provisioning model when a non-default orgId is referenced upfront.

Given that this is a required OSS use case for us, could you please clarify whether this has been evaluated and if there are any plans to support it (or if it is intentionally out of scope)?

Thanks

**Comment by sathieu** [comment]
@yashikapathy28 we made Grafana Org Operator for this use case: https://gitlab.com/kubitus-project/grafana-org-operator

**Comment by yashikapathy28** [comment]
Thanks for the suggestion!

However, we’re specifically looking for native support within Grafana OSS rather than relying on an external operator like the Grafana Org Operator.

Our use case is centered around provisioning in Kubernetes, where we expect Grafana to handle this scenario more gracefully. Ideally:

-  If Grafana OSS cannot create orgIds during provisioning, it should not fail fatally when encountering dashboards or datasources referencing non-existent orgs.
- Instead, provisioning should tolerate such cases (e.g., skip them temporarily).
- Once the corresponding orgs are created (via API or other means), Grafana should automatically apply those previously skipped dashboards/datasources, rather than leaving them permanently ignored.

At the moment, the CrashLoopBackOff behavior when a non-default orgId is referenced makes provisioning workflows quite brittle in Kubernetes environments.

Could you please clarify:

1. Whether this behavior is intentional or considered a limitation?
2. If there are any plans to improve provisioning to handle such cases more gracefully within Grafana itself?

**Comment by spatterIight** [comment]
> 2. If there are any plans to improve provisioning to handle such cases more gracefully within Grafana itself?

See https://github.com/grafana/grafana/pull/78253#issuecomment-1815222389, "We do not want to add more to this system"
